sklearn/utils/testing.py

```diff
@@ -375,6 +375,38 @@ def assert_raise_message(exceptions, message, function, *args, **kwargs):
                              (names, function.__name__))
 
 
+def assert_allclose_dense_sparse(x, y, err_msg=''):
+    """Assert allclose for sparse and dense data.
+
+    Both x and y need to be either sparse or dense, they
+    can't be mixed.
+
+    Parameters
+    ----------
+    x : array-like or sparse matrix
+        First array to compare.
+
+    y : array-like or sparse matrix
+        Second array to compare.
+
+    err_msg : string, default=''
+        Error message to raise.
+    """
+    if sp.sparse.issparse(x):
+        if not sp.sparse.issparse(y):
+            raise ValueError("Can only compare two sparse matrices,"
+                             " not a sparse matrix and an array.")
+        x = x.tocsr()
+        y = y.tocsr()
+        x.eliminate_zeros()
```

Comment:
Hmm... I'm not sure this is necessary or sufficient. What you should have, though, is `sum_duplicates`.